Added stream profiling markers.

MoFtZ opened the pull request:
Fixes #399.

This is a proposed PR to add high resolution timers for Cuda and OpenCL accelerators.

It introduces the concept of a `ProfilingMarker` that can be added to an `AcceleratorStream`.
For OpenCL, the user is required to enable profiling - `contextBuilder.Profiling()`.

Example usage:
```CSharp
using var start = accelerator.DefaultStream.AddProfilingMarker();
launchKernelToBeProfiled();
using var end = accelerator.DefaultStream.AddProfilingMarker();

TimeSpan elapsedTime = end - start; // or end.MeasureFrom(start);
```

NOTE: On the CPU accelerator, it currently uses `DateTime.UtcNow`, which apparently has approximately 15 millisecond accuracy. If this is not acceptable, we will need to investigate a cross-platform high-resolution timer.

NOTE: There appears to be an issue with OpenCL. A workaround was required where the OpenCL event had to be awaited immediately, otherwise the clock timer did not appear to start. Unsure if this is a driver bug on my Intel HD 630.

m4rs-mt commented on Src/ILGPU/Runtime/AcceleratorStream.cs:
```diff
@@ -69,6 +69,12 @@ protected AcceleratorStream(Accelerator accelerator)
+        /// Adds a profiling marker into the stream.
+        /// </summary>
+        /// <returns>The profiling marker.</returns>
+        public abstract ProfilingMarker AddProfilingMarker();
```
Question: Should we also provide an extension/instance method similar to `Synchronize` that adds a marker to the default stream? However, this might be a bad idea as this functionality will only apply to the default stream. @MoFtZ what do you think?

MoFtZ replied:
It probably makes sense to add.

m4rs-mt commented on Src/ILGPU/Runtime/OpenCL/CLAPI.xml:
```diff
@@ -65,7 +65,7 @@
         <Parameter Name="context" Type="IntPtr" />
         <Parameter Name="device" Type="IntPtr" />
-        <Parameter Name="properties" Type="IntPtr" />
+        <Parameter Name="properties" Type="CLCommandQueueProperties" />
```
Question: Is this change correct? If so, was it wrong before?

MoFtZ replied:
It was fine - previously, we always used `IntPtr.Zero`, which is the valid default.

m4rs-mt commented on Src/ILGPU/Runtime/OpenCL/CLAccelerator.cs:
```diff
@@ -318,7 +318,7 @@ protected override CLKernel CreateKernel(CLCompiledKernel compiledKernel)
         /// <inheritdoc/>
         protected override AcceleratorStream CreateStreamInternal() =>
-            new CLStream(this);
+            new CLStream(this, Context.Properties.EnableProfiling);
```
Question: can we check this flag in all cases to ensure consistent behavior for all GPUs?

MoFtZ replied:
OK, will update.

MoFtZ commented:
Updated PR.

Adding a profiling marker without enabling profiling will throw an exception.
Added `AddProfilingMarker` extension method for `Accelerator`.

m4rs-mt commented on Src/ILGPU/Runtime/Cuda/CudaNativeMethods.cs:
```diff
@@ -248,6 +248,37 @@ enum Peer2PeerAttribute
+        CU_EVENT_DEFAULT = 0,
+
+        /// <summary>
+        /// Specifies that the created event should use blocking synchronization. A CPU
```
I am not sure if we are allowed to include the driver API documentation here... I would feel better if we just briefly summarize in one line roughly what the content is 🤓
